### ISSUE TITLE
Feature request MissingMemberHandling.Default

### ISSUE THREAD

**Issue by MichaRotstein** [JamesNK/Newtonsoft.Json#2519]
I suggest adding MissingMemberHandling.Default 

If set and a member is missing during deserialization, try to call a default constructor on the missing type instead of setting it to null.

If there is no default consatructor set it to null (ie fallback to MissingMemberHandling.Ignore

This is especially useful when working with `#nullable enable` - one is not expecting to get null values without compiler warnings in this case.

Thanks

**Comment by sungam3r** [comment]
Breaking change.

**Comment by MichaRotstein** [comment]
@sungam3r  I don't understand, as far as I can see this is not a breaking change. It will only take effect if/when someone is using the suggested new option

**Comment by MichaRotstein** [comment]
@sungam3 fair enough I didn't mean that, how about "TryDefaultConstructor" ? or do you have a better name ?

**Comment by ChrisSchaller** [comment]
What is the business value to this? `Class` (and `Record`) definitions already define the _default_ state of the object that is instantiated before the properties are set during deserialization. 
Serialization and Deserialization should be repeatable, the process should not inject information that was not already there.

Consider the following class:

    public class Student
    {
        public string Name { get;set; }
        public Student OlderSibling { get;set; }
    } 

The serialized state for a single record might look like this

    { "Name": "Test Kid" }

With your `MissingMemberHandling.Default` the deserialized object, if serialized might now look like this:

    { 
        "Name": "Test Kid",
        "OlderSibling": { } 
    }

But now if we deserialize this, we start to get more interesting results, in this case the `Name` field will be initialized to an empty string, because it was missing, and it will have another older sibling:

    { 
        "Name": "Test Kid",
        "OlderSibling": {
            "Name": "",
            "OlderSibling": { }
         } 
    }

Where might this issue present itself? If your application allowed a `Student` to be loaded from a serialized source, then after making changes to _other fields_ they then save this record to a serialized target. Now imagine the user loads and saves the same record many times, suddenly "Test Kid", who is actually an only child in real life, now has a lot of siblings in the application.

Is this a good design to encapsulate a _Student_, not by far... But only the _Author_ of the `Student` class and what ever parent structure this record is referenced from knows what the intent of this design was, the definition of the class itself is an expression of that intent, so it is within the definition of the class that the instructions on how to manage the serialization and deserialization should be defined, or in the business logic that manages the serialization process.

Right now, if you actually _wanted_ this behaviour, you would define your class with an initializer for the property that you want to be pre-initialized, noting that this only works if `MissingMemberHandling.Ignore` is set:

    public class Student
    {
        public string Name { get;set; }
        public Student OlderSibling { get;set; } = new Student();
    } 

> you could also define this in the default constructor for the class, IMO that is a habit that leads to maintenance issues down the track and is harder to infer from the model during design time when you are using this class.

In this fictitious class, it might be acceptable for `Name`  to be pre-init, but not `OlderSibling`, but a `MissingMemberHandling.Default` would not be able to differentiate between different members, it is all or none.

A more acceptable `Student` class might be this:

    public class Student
    {
        public string Name { get;set; } = String.Empty;
        public Student OlderSibling { get;set; }
    } 

Now the deserializing an empty JSON object, and then serializing would look like this:

    { "Name": "" }

If you wanted specifically to manage deserialization for specific properties or types, you could always use one of the deserialization call backs:

    public class Student
    {
        public string Name { get;set; }
        public Student OlderSibling { get;set; }
        
        [OnDeserializing]
        internal void OnDeserializingMethod(StreamingContext context)
        {
            Name = "Default Student";
        }
    } 

Then run the following code:

    var json = @"[
        { },
        { ""Name"": """" },
        { ""Name"": ""Little Johnny"" }
    ]";

    var students = Newtonsoft.Json.JsonConvert.DeserializeObject<Student[]>(json, new JsonSerializerSettings 
                        { 
                            MissingMemberHandling = MissingMemberHandling.Ignore 
                        }
                    );

    Console.WriteLine(JsonConvert.SerializeObject(students, Formatting.Indented, new JsonSerializerSettings 
                        { 
                            MissingMemberHandling = MissingMemberHandling.Ignore, 
                            NullValueHandling = NullValueHandling.Ignore 
                        }
                    ));

Which should output this:
```json
[
  {
    "Name": "Default Student"
  },
  {
    "Name": ""
  },
  {
    "Name": "Little Johnny"
  }
]
```

If you do not want to, or cannot manipulate the class definition, then you could always resort to a `JsonConverter`.
The following is an example implementing the `CustomCreationConverter`:

    public class StudentDTO
    {
        public string Name { get; set; }
        public Student OlderSibling { get; set; }
    }
    public class StudentConverter : CustomCreationConverter<StudentDTO>
    {
        public override StudentDTO Create(Type objectType)
        {
            return new StudentDTO { Name = "Custom Creation" };
        }
    }

Execution:

    json = @"[
        { },
        { ""Name"": """" },
        { ""Name"": ""Little Johnny"" }
    ]";

    var students = Newtonsoft.Json.JsonConvert.DeserializeObject<StudentDTO[]>(json,new StudentConverter() );

    Console.WriteLine(JsonConvert.SerializeObject(students, Formatting.Indented, new JsonSerializerSettings 
                        { 
                            MissingMemberHandling = MissingMemberHandling.Ignore, 
                            NullValueHandling = NullValueHandling.Ignore 
                        }
                    ));

output:

```json
[
  {
    "Name": "Custom Creation"
  },
  {
    "Name": ""
  },
  {
    "Name": "Little Johnny"
  }
]
```

**Comment by kofifus** [comment]
Thanks @ChrisSchaller , what you wrote above amounts to 'in some scenarios using this option may give unwanted results' sure I agree, in those cases (ie self referencing records) that option should not be used.

My use case:

C# 9 introduced the short record definition command _and_ non nullable reference types.
It is now common to right:

```
#nullable enable

record Rec(
   int Age, 
   Employee Employee
   .. many other fields ..
);
```

This form without a constructor is very advantageous. It removes the need to remember to correctly edit the ctor when adding or removing fields which both tedious and a common source of bugs.

working with `#nullable enable` one expects these record fields to **not be null**. This is crucial.

To have such a record de serialized and have a field set to null even after explicitly specifying `#nullable enable` is very confusing - I would consider it a json.net bug

There are three ways out here
 - as you suggest to forgo the benefits of not having to write a ctor and write it explicitly with default parameters where needed
 - use a custom converter etc this is both tedious and has the exact same problems with self references
 - my request for a specific directive to `DeserializeObject` to tell it in this case I want to use the empty ctor for missing fields. I think this will be a useful addition
 
 I hope this is clearer

**Comment by ChrisSchaller** [comment]
The point of `#nullable enable` is not to make sure that nothing in your code is null, it is to deliberately force you to check that values are null. There is no such expectation that the record fields cannot be null. only that if a record has an immutable property that is instantiated as null, then it's value cannot change.

If your entire code base cannot be null, then we would have to check for _zeros_ or _defaults_ for every comparison, which would be almost as bad. Even with records, I cannot see the argument for setting the `Employee` property to a default instance.

**Comment by ChrisSchaller** [comment]
I'll leave you alone after this :) the main issue for me is that if a review all my recent projects, that cover finaces, Logistics, IOT business domins, I've even deployed a few micro-service and bot architectures, the amount of individual classes, out of the thousands from these real world business data model designs that could legitimately use this feature, where _ALL_ properties _could_ and _should_ be pre-initialised to non-null values is very close to zero. Its pretty much only meta-data structures or single purpose classes where this could be applied.

I would instead argue that having a `MissingMemberHandling.Null` option would have more value, being able to deliberately override the default initialization to a null state (for nullable types). In fact it would make more sense to me if your name of `MissingMemberHandling.Default` actually set the value to `null` as that is the _default_ value for a nullable type. From an implementation point of view this would be the same as using the `default` literal or the generic `default(T)`.

**Comment by MichaRotstein** [comment]
yes forget about the name it's confusing I now changed the name in the request to TryDefaultConstructor

As for your projects .. are you using records for those thousand of cases ? are you working with #nullable enabled ? otherwise they are not relevant here. 

> Its pretty much only meta-data structures or single purpose classes where this could be applied

this could be applied to almost any (except self referencial etc which is rare) records

**Comment by ChrisSchaller** [comment]
Even when using _records_, `null` is still a significant value, with `#nullable enabled` you just use the null coalescing operator or deliberately check for null.

> This type of behaviour, to instantiate a property with a default value that is NOT defined as part of the property declaration or meta-data in the class definition is _unusual_ and does not conform to any existing or previous framework standards.
>> This perhaps should be a request of `record` type implementation itself, and not specifically anything to do with JSON.Net? 

When using records, it's still simple syntax to declare initialization, the main argument is that only the class or record definition itself can "know" if it is valid for that property to be in a null state or not.

> there may be some value in adding this as an option to `JsonObjectAttribute` but as per this #2380 this attribute is ignored if the property was not declared in the source text that is being deserialized.

It would not be hard to write a generic converter to inspect property attributes on the target class and use them when initialising the initial state of the object instance, however you might have to work a bit harder to make it suit immutable properties on a `record` when the property is missing from the source text, specifically in the context of records, there might be an argument, but the accepted solution is either to use the verbose record declaration. 

_Again, this is more of an issue with `record` declaration syntax and less about deserialization._


### Regarding `#nullable enabled`:

Even if the serializer _did_ generate a default instance of the object to inject into the property you will still be required to use null checking because the property type is itself nullable. To get around this you should use a non-nullable type. 

If the property is itself a `record` type, then what would be the purpose of a value that cannot be changed, it will always be the default state. In this state, I think that null has much more value, rather than having to determine if the value is null because it was not initialized correctly, or if it was deliberately initialized with that effectively _empty_ value.

> This discussion goes full circle, your example and reasoning highlight why it is important that as part of serialization we should be able to _force_ these properties to be `null`. These values did not exist in the serialized content, and therefore there is a strong argument that they should not exist in the hydrated object graph after deserialization either.

**Comment by MichaRotstein** [comment]
This is becoming off topic, to me you misunderstand non-nullable reference types but let's leave it that. You don't think my feature suggestion makes sense I hope others will.
